Close #1013: Add permalink to checklist items

andrewborstein opened the pull request:
Closes #1013

#### Permalink Example

Visitors on [the Checklist page](https://www.a11yproject.com/checklist/) can now 
1. Expose the URL to an individual checklist item by pressing its associated "Permalink" anchor element
2. Visit a checklist item permalink URL like `/checklist/#validate-your-html`, which scrolls the item into view and (when JS is enabled) opens its details

<img width="1131" alt="a11yproject.com checklist item opened and featuring new permalink" src="https://user-images.githubusercontent.com/7654254/89115457-5665a780-d456-11ea-921e-cd3c9ee0b978.png">
 
#### Notes and Questions
- I'm open to other UI suggestions! I wanted to aim for something simple, conventional, and matching existing styles. But I'm unsure if this link could benefit from some other visual distinction, particularly from the checklist item citation link.
- Is "Permalink" too jargony? I also considered copy like "Share Link" or "Shareable URL". 
- Does associating the link to its checklist item via `aria-describedby` sufficiently meet [2.4.4 Link Purpose (In Context)](https://www.w3.org/TR/WCAG20/#navigation-mechanisms-refs)? Including the entire title in the visible link text seemed liked overkill and rendering visually hidden text seemed unnecessary when the text already exists and could be linked using this technique. 
- I noticed a bug in Chrome where continually refreshing the page with a permalink URL in place continues scrolling further and further down the page instead of reloading the same scroll position. The issue doesn't occur when visiting the URL in new tabs or windows, or in Firefox or Safari. [It might be related to this bug 🤷 ](https://support.google.com/chrome/thread/11993079?hl=en), but doesn't seem like a critical user flow that needs to be supported above and beyond the first page visit always working.
- I wanted to extract `checklist.title | slugify` to a local variable but couldn't figure out how. I don't have a ton of Nunjucks / 11ty experience, so insight there is appreciated!

ericwbailey commented:
Awesome, thank you for doing this. To answer your questions: 

> I'm open to other UI suggestions! I wanted to aim for something simple, conventional, and matching existing styles. But I'm unsure if this link could benefit from some other visual distinction, particularly from the checklist item citation link.

I like this approach. Simple, direct and to the point.

> Is "Permalink" too jargony? I also considered copy like "Share Link" or "Shareable URL".

I'm always of mixed opinions about the term permalink, but I think "Share Link" would probably be the way to go (especially if the checklist it's name ever updates 😅).

>Does associating the link to its checklist item via `aria-describedby` sufficiently meet [2.4.4 Link Purpose (In Context)](https://www.w3.org/TR/WCAG20/#navigation-mechanisms-refs)? Including the entire title in the visible link text seemed liked overkill and rendering visually hidden text seemed unnecessary when the text already exists and could be linked using this technique.

I'd actually like to use visually hidden text. I'm thinking both A) First rule of ARIA and B) I'm unsure how good the support would be for AT that generates a list of generated links. Mind updating it to use a visally hidden class?

>I wanted to extract `checklist.title | slugify` to a local variable but couldn't figure out how. I don't have a ton of Nunjucks / 11ty experience, so insight there is appreciated!

I believe slugify only works in Eleventy's loop through [Collections](https://www.11ty.dev/docs/collections/), but I can play around and see what I can do if you want.

andrewborstein commented:
Great points @ericwbailey! Thanks so much for the feedback. 

I changed "Permalink" to "Share Link" and removed the aria in favor of visually hidden text. I also found some documentation on [the `set` tag in Nunjucks](https://mozilla.github.io/nunjucks/templating.html#set), which I used to create local variables.

#### Updated "Share Link" Example

<img width="1102" alt="A11y Project checklist item opened and featuring new share link" src="https://user-images.githubusercontent.com/7654254/89133731-5ca66300-d4ec-11ea-8a50-2d79f82719fe.png">

When the example link is encountered by AT its accessible name should read, "Share Link to checklist item: Validate your HTML".

ericwbailey commented:
Love it. Thank you so much for doing this!
